includes/Core/Authentication/Authentication.php

```diff
@@ -443,6 +451,47 @@ public function get_disconnect_url() {
 		);
 	}
 
+	/**
+	 * Gets the proxy permission url
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return string Proxy permission url.
+	 */
+	private function get_proxy_permissions_url() {
+		return add_query_arg(
+			array(
+				'action' => Google_Proxy::ACTION_PERMISSIONS,
+				'nonce'  => wp_create_nonce( Google_Proxy::ACTION_PERMISSIONS ),
+			),
+			admin_url( 'index.php' )
+		);
+	}
+
+	/**
+	 * Handles proxy permissions
+	 *
+	 * @since n.e.x.t
+	 */
+	private function handle_proxy_permissions() {
+		if ( ! current_user_can( Permissions::AUTHENTICATE ) ) {
+			wp_die( esc_html__( 'You need a higher level of permission.', 'google-site-kit' ) );
+		}
+
+		if ( ! $this->credentials->using_proxy() ) {
+			wp_die( esc_html__( 'You need a higher level of permission.', 'google-site-kit' ) );
+		}
+
+		$nonce = $this->context->input()->filter( INPUT_GET, 'nonce' );
+		if ( ! wp_verify_nonce( $nonce, Google_Proxy::ACTION_PERMISSIONS ) ) {
```

Comment:
Can we move the nonce check up to be the first check? That's the most basic one - if the nonce is invalid, clearly the request is not authorized. This is also how the setup handler works.